### PR TITLE
HeaderWriterFilter writes headers at beginning

### PR DESCRIPTION
**Pull request by ankurpathak**
Add support for HeaderWriterFilter to write headers at the beginning of the request

Fixes: gh-6501

### CODE REVIEW

**Comment by ankurpathak** [comment]
@jzheaux I am not getting any Idea for tests. Please give some idea how I can write test.

File: web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java [hunk 1]
```diff
@@ -73,12 +78,28 @@ protected void doFilterInternal(HttpServletRequest request,
...
 	}
 
+	/**
+	 * Allow writing headers at the beginning of the request.
+	 *
+	 * @param shouldWriteHeadersEagerly boolean to allow writing headers at the beginning of the request.
+	 * @author Ankur Pathak
+	 * @since 5.2
+	 */
+	public void setShouldWriteHeadersEagerly(boolean shouldWriteHeadersEagerly) {
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: Let's go ahead and add a couple of tests that confirm that this functionality works. 

One test in `HeaderWriterFilterTests` that confirms that when using the setter, then the headers are written early.

And one in `HeadersConfigurerTests` that confirms that the setting can be made via an `ObjectPostProcessor`. (If you haven't done that before, you can see an example in the ticket description.) In this case, I think you can just confirm that the headers still get written.

Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]:
```
http
    .addObjectPostProcessor(new ObjectPostProcessor<HeaderWriterFilter>() {
        public HeaderW riterFilter postProcess(HeaderWriterFilter filter) {
            filter.setShouldWriteHeadersEagerly(true);
            return filter;
        }
    }
```
I can not find any addObjectPostProcessor method in HttpSecurity. Please check this url:
https://docs.spring.io/spring-security/site/docs/5.2.0.BUILD-SNAPSHOT/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html

Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]:
```@EnableWebSecurity
	static class HeadersAtTheBeginningOfRequestConfig extends WebSecurityConfigurerAdapter {
        @Override
        protected void configure(HttpSecurity http) throws Exception {
            http
                    .headers()
                    .addObjectPostProcessor(new ObjectPostProcessor<HeaderWriterFilter>() {
						@Override
						HeaderWriterFilter postProcess(HeaderWriterFilter filter) {
							filter.shouldWriteHeadersEagerly = true
							return filter
						}
					})
        }
    }
```
This inside groovy is giving me errors below:
screenshot from 2019-02-08 09-11-54

Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: I am done with changes.

File: web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java [hunk 2]
```diff
@@ -73,12 +78,28 @@ protected void doFilterInternal(HttpServletRequest request,
 				headerWriterResponse);
 
 		try {
+			if (shouldWriteHeadersEagerly) {
+				headerWriterResponse.writeHeaders();
```
[the remark this answers is not in the document]
Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: Done.

File: web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java (hunk 1 above)
Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: > Let's go ahead and add a couple of tests that confirm that this functionality works.
> 
> One test in `HeaderWriterFilterTests` that confirms that when using the setter, then the headers are written early.
> 
> And one in `HeadersConfigurerTests` that confirms that the setting can be made via an `ObjectPostProcessor`. (If you haven't done that before, you can see an example in the ticket description.) In this case, I think you can just confirm that the headers still get written.

Done.

Comment (jzheaux) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: Looks like you found it, yes, the correct way is:

```java
http
    .headers()
        .withObjectPostProcessor(...
```

Comment (ankurpathak) [reply web/src/main/java/org/springframework/security/web/header/HeaderWriterFilter.java]: @jzheaux You can review it. Its complete.

File: config/src/test/groovy/org/springframework/security/config/annotation/web/configurers/HeadersConfigurerTests.groovy [hunk 3]
```diff
@@ -565,4 +567,20 @@ class HeadersConfigurerTests extends BaseSpringSpec {
 		    responseHeaders == ['Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains ; preload']
 	}
 
+    def "headers written at beginning of request"() {
```
Comment (jzheaux) [inline config/src/test/groovy/org/springframework/security/config/annotation/web/configurers/HeadersConfigurerTests.groovy]: Please make this a Java test. We are in the process of coverting the Groovy tests over to Java, so we don't want to make more work for ourselves.

File: config/src/test/java/org/springframework/security/config/annotation/authentication/configurers/HeadersConfigurerTestsConfigs.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Tests config for {@link HeadersConfigurerTests}.
+ *
+ * @author Ankur Pathak
+ */
+public class HeadersConfigurerTestsConfigs {
+
+	@EnableWebSecurity
+	public static class HeadersAtTheBeginningOfRequestConfig extends WebSecurityConfigurerAdapter {
```
Comment (jzheaux) [inline config/src/test/java/org/springframework/security/config/annotation/authentication/configurers/HeadersConfigurerTestsConfigs.java]: Usually, we place configs alongside the tests, as inner classes. Could we do that here, too, please?

File: config/src/test/groovy/org/springframework/security/config/annotation/web/configurers/HeadersConfigurerTests.groovy (hunk 3 above)
Comment (ankurpathak) [reply config/src/test/groovy/org/springframework/security/config/annotation/web/configurers/HeadersConfigurerTests.groovy]: Done.

File: config/src/test/java/org/springframework/security/config/annotation/authentication/configurers/HeadersConfigurerTestsConfigs.java (hunk 4 above)
Comment (ankurpathak) [reply config/src/test/java/org/springframework/security/config/annotation/authentication/configurers/HeadersConfigurerTestsConfigs.java]: Done.

**Comment by jzheaux** [comment]
@ankurpathak Thanks for the PR! This is now merged into `master` via ac13b55ecd93b9eaf40866aabb043b14dd4f3350. 

Also, I added a polish commit of 2b960b074b8518c81a789214e542954d82deedb8 for the Java Config test - please take a look at the commit message for future reference.

**Pull request closed without merging**